### PR TITLE
[New] `consistent-type-specifier-style`: add rule

### LINKED ISSUE #2469: New rule: consistent inline type specifier usage
In both flow and TS you can mark an import as a type (or typeof for flow) via a specifier.
Both languages support two forms:

top-level, which marks all names in the import as type-only:
```ts
import type Foo from 'Foo';
import type {Bar} from 'Bar';
// ts only 
import type * as Bam from 'Bam';
// flow only
import typeof Baz from 'Baz';
```

inline, which marks just one name in the import as type-only:
```ts
import {type Foo} from 'Foo';
// flow only
import {typeof Bar} from 'Bar';
```

Users care about consistency when it comes to the use of inline type imports - they either want them, or they don't.
I would like to propose a rule which enforces either for, or against the use of inline type specifiers.

The rule would have an option which would enforce one way or another, i.e.
```ts
type Options = [
  'perfer-inline' | 'prefer-top-level',
];
```

### ✅ Valid always

```ts
import type Foo from 'Foo'; // default specifiers are always good
import type * as NS from 'Bar'; // namespace specifiers are always good
```

### ✅ Valid with `prefer-top-level`, but ❌ Invalid with `prefer-inline`

```ts
import type {Foo} from 'Foo';
import type Foo, {Bar} from 'Foo';
// flow only
import typeof {Foo} from 'Foo';
```

### ❌ Invalid with `prefer-top-level`, but ✅ Valid with `prefer-inline`

```ts
import {type Foo} from 'Foo';
import Foo, {type Bar} from 'Foo';
// flow only
import {typeof Foo} from 'Foo';
```

## Fixer

For `prefer-top-level`:
```ts
import {Foo, type Bar, typeof Baz} from 'Foo';
// becomes
import {Foo} from 'Foo';
import type {Bar} from 'Bar';
import typeof {Baz} from 'Baz';
```

For `prefer-inline`:
```ts
import type {Foo, Bar} from 'Foo';
// becomes
import {type Foo, type Bar} from 'Foo';

import type Foo, {Bar} from 'Foo';
// becomes
import type Foo from 'Foo';
import {type Bar} from 'Foo';
```

Note that the fixer need not necessarily concern itself with de-duplicating the imports, as `import/no-duplicates` can be turned on by the user to handle that concern.

### PR DESCRIPTION
**Pull request by bradzacher**
Fixes #2469

Adds a new rule as specced in #2469
To test this I had to bump the TS version to 4.5, as that was the first version that supported inline type specifiers.

### CODE REVIEW

**Comment by bradzacher** [comment]
image

big oof - testing against all of these old versions of things is really tough to get right

File: docs/rules/consistent-type-specifier-style.md
Comment (snewcomer) [inline docs/rules/consistent-type-specifier-style.md:30]: support for `inlineTypeImport` https://github.com/import-js/eslint-plugin-import/pull/2475

File: tests/src/rules/consistent-type-specifier-style.js
Comment (bradzacher) [reply tests/src/rules/consistent-type-specifier-style.js:83]: ohh that's a really good test case I forgot about.
this one is even better though:
```ts
import AValue, { type Foo } from 'Foo';
```
Because it should fix to
```ts
import AValue from 'Foo';
import type Foo from 'Foo';
```
i.e. we have to remove the empty `{}` as well

I wonder... @ljharb - do you have a preference here?
Should there be a new rule which enforces against empty named specifier blocks, or should we just bring that complexity into each lint rule?

Comment (ljharb) [reply tests/src/rules/consistent-type-specifier-style.js:83]: @bradzacher Isn't that invalid syntax? `import {} from` is already invalid.

If it's valid to have `import type {} from` then i think we just need to handle that in individual rules.

(Sorry for the delayed response; i'd like to land this soon so I'll keep a closer eye on it)

Comment (bradzacher) [reply tests/src/rules/consistent-type-specifier-style.js:83]: > Isn't that invalid syntax? import {} from is already invalid

According to [the spec](https://262.ecma-international.org/6.0/#sec-imports) `{}` is allowed:

```
NamedImports :
    { }
    { ImportsList }
    { ImportsList , }
```

Based on testing - all eslint parsers allow it.
Also the flow/hermes/typescript/babel parsers allow `import type {}`

Comment (ljharb) [reply tests/src/rules/consistent-type-specifier-style.js:83]: oh oof, gross. ok then i think a) we'd need a separate rule to ban empty named import/export blocks, and b) this rule's autofixer should just always be removing empty blocks when removing duplicates.

Comment (bradzacher) [reply tests/src/rules/consistent-type-specifier-style.js:83]: there is *some* additional complexity that will get added to the fixer for this, but we can definitely do it!

The other option is to let this fixer purely focus on moving specifiers around, and then let the new `no-empty-named-blocks` focus on cleaning up the empty curlies - similar to how the fixer right now does not attempt to consolidate imports, and just moves things around.

Comment (ljharb) [reply tests/src/rules/consistent-type-specifier-style.js:83]: I suppose that's an option too - but I think in the common case people will not want empty braces, and it's nice when autofixers optimize for the common case.

That said, certainly if someone _wants_ empty braces for some reason, that's only possible if this autofixer leaves them behind :-/

File: src/rules/consistent-type-specifier-style.js
Comment (ljharb) [inline src/rules/consistent-type-specifier-style.js:55]:
```suggestion
            fix(fixer) {
              const kindToken = sourceCode.getFirstToken(node, { skip: 1 });

              return [].concat(
                kindToken ? fixer.remove(kindToken) : [],
                node.specifiers.map((specifier) => fixer.insertTextBefore(specifier, `${node.importKind} `),
              );
            }
```

File: src/rules/consistent-type-specifier-style.js
Comment (ljharb) [inline src/rules/consistent-type-specifier-style.js:135]:
```suggestion
                function* removeSpecifiers(specifiers) {
```

**Review by ljharb — COMMENTED** [review]
I'd prefer to avoid use of generators and loops, but that's also something I can clean up once the tests are passing.

How does a Flow user who isn't using the inline `typeof` syntax already, autofix to `import typeof`?

I think that if we can get this to leave behind the empty braces, we'll be OK pending the other comments. I'd love to land this soon.

**Comment by bradzacher** [comment]
Now that the TS 4.8 release is done, I should be able to get back to this this week (hopefully).

**Comment by ljharb** [comment]
I'm holding off on the next release for this :-) looking forward to you getting back to it.

**Comment by bradzacher** [comment]
assuming test pass - this should be good to go now.
I updated the fixer to (a) cleanup unnecessary commas and (b) cleanup empty named blocks

File: src/rules/consistent-type-specifier-style.js
Comment (ljharb) [inline src/rules/consistent-type-specifier-style.js:39]: can this line be covered by a test case?

File: src/rules/consistent-type-specifier-style.js
Comment (ljharb) [inline src/rules/consistent-type-specifier-style.js:94]: also this one?

File: src/rules/consistent-type-specifier-style.js
Comment (ljharb) [inline src/rules/consistent-type-specifier-style.js:226]: let's move this to the top of the module, so we aren't relying on hoisting

**Review by ljharb — COMMENTED** [review]
Looks great - have these items been addressed?
> How does a Flow user who isn't using the inline typeof syntax already, autofix to import typeof?

> leave behind the empty braces

**Comment by bradzacher** [comment]
> leave behind the empty braces

The fixer should no longer leave behind empty braces.

> How does a Flow user who isn't using the inline typeof syntax already, autofix to import typeof?

If I'm understanding your question correctly:

- If they don't want to use the inline specifiers at all, then they can use the `prefer-top-level` setting.
- If they want to use inline `type` with top-level `typeof`, the rule does not currently have handling for this (could be a future enhancement if people want it!).

File: src/rules/consistent-type-specifier-style.js
Comment (bradzacher) [reply src/rules/consistent-type-specifier-style.js:94]: We could actually delete this check entirely.
It's covered above by the checks on line 80-84.
It's a syntax error to have a namespace specifier with any other kind of specifier.

This is just me doing my due diligence because I'm used to covering all of the bases due to TS types.

Would you like me to delete it?

File: src/rules/consistent-type-specifier-style.js
Comment (bradzacher) [reply src/rules/consistent-type-specifier-style.js:44]: This should already be covered by:
- https://github.com/bradzacher/eslint-plugin-import/blob/2df9be6b7edd150c9ac94b2d479ea51f5fbe66d0/tests/src/rules/consistent-type-specifier-style.js#L16-L19
- https://github.com/bradzacher/eslint-plugin-import/blob/2df9be6b7edd150c9ac94b2d479ea51f5fbe66d0/tests/src/rules/consistent-type-specifier-style.js#L40-L43
- https://github.com/bradzacher/eslint-plugin-import/blob/2df9be6b7edd150c9ac94b2d479ea51f5fbe66d0/tests/src/rules/consistent-type-specifier-style.js#L207

and a few others

Comment (ljharb) [reply src/rules/consistent-type-specifier-style.js:55]: if it were, then i wouldn't expect to see the "no coverage" annotation

File: src/rules/consistent-type-specifier-style.js
Comment (ljharb) [reply src/rules/consistent-type-specifier-style.js:104]: if it's not used, then it's probably best to delete it.

**Comment by ljharb** [comment]
Rebased; looks like this is good to go, pending any "uncovered line" annotations (https://github.com/import-js/eslint-plugin-import/pull/2473#discussion_r963255391 and https://github.com/import-js/eslint-plugin-import/pull/2473#discussion_r963255298)

File: src/rules/consistent-type-specifier-style.js
Comment (bradzacher) [reply src/rules/consistent-type-specifier-style.js:55]: Weird - I'm not sure what's happening with the coverage report.
Locally I can see it's getting hit in the tests:
image

**Review by ljharb — APPROVED** [review]

**Comment by latin-1** [comment]
https://github.com/microsoft/TypeScript/issues/47118

```ts
import type { X } from "...";
```

```ts
import { type X } from "...";
```

These two code snippets behave differently. Perhaps we should add a special treatment for this.

**Comment by ljharb** [comment]
It's not clear what the difference is from that issue. Can you elaborate?

**Comment by latin-1** [comment]
https://github.com/typescript-eslint/typescript-eslint/issues/6338
I posted more details here. I'm not sure which plugin this should be handled in.
It's a mess. We may need two or more rounds of code fixes to make it work.

**Comment by ljharb** [comment]
eslint on the CLI runs multiple passes, so it's only in an editor that you're likely to run into this problem.

**Pull request merged**
